lib/models.py

```diff
@@ -2,13 +2,231 @@
 import openai
 from typing import Any, Callable, List, Optional
 
+import numpy as np
+import itertools
+
 class CoT():
     """
     Creates a CoT wrapper around an LLM. 
 
     Is called exactly like an LLM, just excecutes a chain of thought to get answer.
+
+    Based on code from this paper: https://arxiv.org/pdf/2305.10601.pdf
     """
-    pass
+
+    num_steps: int = 3               # the depth of the chain of thought tree
+    num_select_sample: int = 2       # how many samples to keep exploring at the level of the tree
+    num_generate_samples: int = 3    # how many samples to generate for each sample of the tree (branching fraction)
+    verbose: bool = False            # weather or not to print out the CoT
+    selection_method: str = 'greedy' # Describes how set of best possible solutions is selected. Can either be 'greedy' or 'sample'.
+    max_num_tries: int = 3           # maximum number of tries to get a score or final answer before it gives up.
+    return_boolean: bool = True      # flag which, when set to true, forces CoT to output a boolean string (either "true" or "false")
+    get_samples_prompt: str = ''     # prompt to get the samples
```

Comment:
Add to the comment which format string specifiers are required, e.g., for samples_prompt, “Problem” and “Previous_CoT”